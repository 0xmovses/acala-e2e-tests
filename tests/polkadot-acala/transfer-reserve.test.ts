import { afterAll, describe, expect, it } from "vitest";
import { connectVertical } from "@acala-network/chopsticks";

import { expectEvent, matchSnapshot, setupContext, testingPairs } from "../helper";

describe("Polkadot <-> Acala", async () => {
  const polkadot = await setupContext({
    endpoint: "wss://rpc.polkadot.io",
  });
  const acala = await setupContext({
    endpoint: "wss://acala-rpc-1.aca-api.network",
  });
  await connectVertical(polkadot.chain, acala.chain);

  const { alice } = testingPairs();

  afterAll(async () => {
    await polkadot.teardown();
    await acala.teardown();
  });

  it("Polkadot transfer assets to Acala", async () => {
    await polkadot.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
    });

    await matchSnapshot(polkadot.api.query.system.account(alice.address));
    await matchSnapshot(acala.api.query.tokens.accounts(alice.address, { token: "DOT" }));

    await polkadot.api.tx.xcmPallet
      .reserveTransferAssets(
        {
          V0: {
            X1: {
              Parachain: 2000,
            },
          },
        },
        {
          V0: {
            X1: {
              AccountId32: {
                network: "Any",
                id: alice.addressRaw,
              },
            },
          },
        },
        {
          V0: [
            {
              ConcreteFungible: { id: "Null", amount: 100e10 },
            },
          ],
        },
        0
      )
      .signAndSend(alice);

    await polkadot.chain.newBlock();
    await acala.chain.upcomingBlock();

    await matchSnapshot(polkadot.api.query.system.account(alice.address));
    await expectEvent(polkadot.api.query.system.events(), {
      event: expect.objectContaining({
        section: "xcmPallet",
        method: "Attempted",
      }),
    });

    await matchSnapshot(acala.api.query.tokens.accounts(alice.address, { token: "DOT" }));
    await expectEvent(acala.api.query.system.events(), {
      event: expect.objectContaining({
        section: "parachainSystem",
        method: "DownwardMessagesReceived",
      }),
    });
  });
});
