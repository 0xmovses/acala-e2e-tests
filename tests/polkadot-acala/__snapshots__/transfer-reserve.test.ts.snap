// Vitest Snapshot v1

exports[`Polkadot <-> Acala > Polkadot transfer assets to Acala 1`] = `
{
  "consumers": "0",
  "data": {
    "feeFrozen": "0",
    "free": "10,000,000,000,000",
    "miscFrozen": "0",
    "reserved": "0",
  },
  "nonce": "0",
  "providers": "0",
  "sufficients": "0",
}
`;

exports[`Polkadot <-> Acala > Polkadot transfer assets to Acala 2`] = `
{
  "free": "0",
  "frozen": "0",
  "reserved": "0",
}
`;

exports[`Polkadot <-> Acala > Polkadot transfer assets to Acala 3`] = `
{
  "consumers": "0",
  "data": {
    "feeFrozen": "0",
    "free": "8,999,815,091,021",
    "miscFrozen": "0",
    "reserved": "0",
  },
  "nonce": "1",
  "providers": "0",
  "sufficients": "0",
}
`;

exports[`Polkadot <-> Acala > Polkadot transfer assets to Acala 4`] = `
{
  "free": "999,998,266,092",
  "frozen": "0",
  "reserved": "0",
}
`;
